src/client/pfs/pfs.proto

```diff
@@ -253,6 +266,10 @@ message FinishCommitRequest {
   bool empty = 4;
 }
 
+message MergeRequest {
```

Comment:
To match with our existing requests this should probably be called `MergeCommitRequest` since it creates a commit. Also shouldn't there be a corresponding `MergeCommit` rpc in this file? It looks like you're implementing the method on the `apiServer` but it isn't part of the interface.

Reply:
Yeah, I am thinking about backing out on exposing a merge api endpoint, at least for now. I think it is a little unclear what that means in terms of the level based compaction right now. The way I think things should work for now is to have finish commit always do a compaction and for get file to pull from the levels. Also, I think this would be a bit cleaner because the management of a commit's storage would be entirely in pfs.

Reply:
Just to clarify a bit more, what I think this would mean in terms of job output is that workers would write their datum set file sets to paths like `/repo/commit/datumset-id/`, then when finish commit is called, pfs would compact everything under the prefix `/repo/commit/`.

Reply:
I think it's a really good idea to expose a merge rpc. It's a frequently asked for feature so it would be nice to be able to expose it to users, but more importantly it seems like it's a much more tractable problem than larger the outputs from workers, since it's just merging 2 trees, and it should be much easier to test, since you can do it entirely through the pfs api, which means you don't have to worry about where workers write their outputs or anything like that.

Reply:
I'm not really following what you are saying here. I agree we want a merge api endpoint exposed at some point, but I don't think it needs to be available day one when the new storage layer is released. As far as job output is concerned, I guess you are saying something to the effect that workers can just write anywhere and then the master can call merge on those paths? I think a cleaner solution is to have well defined locations for worker output that can be exposed by providing an additional identifier when making put file calls (to specify the datum set id part of the prefix mentioned above). This would essentially mean that upon job completion, the paths under the commit prefix represent the full state of the commit (no extra coordination about object storage paths required). So, we would compact these with the distributed merge. The way I think an exposed merge would work is that you would give it a set of commits and it would merge/compact the compacted file sets for those commits (the level based structure). The outputs of workers would just be temporary file sets that would not be in the level based structure (doing this gives us an advantage in that we can do the least number of intermediary merges to get the final output).

Reply:
I think we should have an implementation of a Merge rpc before we have an implementation of the worker merge, and definitely way before the new storage layer is released. A merge rpc that just merges commits is a lot simpler than the worker merge implementation. Having it and being able to test its semantics will let us be a lot more confident that the semantics are correct. It probably shouldn't even be a distributed merge, at least not at first.

Reply:
I don't think we are on the same page. If the notion of merging/compaction happens behind the pfs layer, why would we need to expose a merge api endpoint to get the expected semantics? Let me just layout how this would work:

For input repos:
```
1. Client calls start commit.
2. Sequence of put files.
3. Client calls finish commit.
  - Pachd instance that finish commit was called on shards the commit's file set (merge the index parts of the commit created through the put file calls and create shards)
  - Merge workers claim shards and merge (these workers are all in pfs, they are not pipeline workers)
  - Pachd instance that finish commit was called on merges/compacts the final results.
```

For output repos (job output):
```
1. Workers call put file for the output commit with an additional identifier associated with the datum set being processed (this extra identifier may not be necessary though).
2. Master pipeline worker calls finish commit which functions exactly the same as finish commit above.
```

There is no notion of merging as far as the pipeline workers are concerned, it all happens in pfs. Exposing a merge api endpoint just adds a merging feature that is not necessary to support pachyderm's fundamental semantics (commit has the merge of all of the content added/deleted in the ancestor commits). An exposed merge feature would use most of the same implementation, but would need some notion of how to handle merge conflicts and checking for a shared root.

I currently have testing for the merge in both the storage layer and pfs layer (pretty basic tests though). The storage layer test is testing the merge implementation in the storage layer (take a set of paths and merge them into an output path). The pfs layer test is testing both the end to end process and the distributed merge (the distributed portion has its own test in the work package, so this test is more of an integration test). This means doing a set of commits with each having a set of put files. Each of these commits are merged during finish commit using the distributed merge process.

Reply:
I see, this makes sense, and seems to imply that we basically already do have merge commit, it's just a recursive copy file from the commit you want to merge, into the commit you want to merge it into. The difference now is that at least part of this process is now distributed. With that in mind I'm cool with not having this `MergeCommit` as part of this PR, although I think it's pretty trivial to add since it's just an alias. We should definitely have it as part of the new storage layer release, since I think it's really easy to do.

Reply:
I'm not really understanding the point of the recursive copy file analogy, but I think we are both fine with not exposing a merge api endpoint for now (at least for this PR). I don't think it is just an alias because of the points I mention above (merge conflicts and shared root). Across repos these might not be a problem, but across branches in the same repo it becomes a concern because of the potential for duplicating content.